src/rust/ide/lib/view/src/graph_editor/component/node/port/output.rs

```diff
@@ -659,26 +733,43 @@ impl OutputPorts {
 
         let port_frp = PortFrp {mouse_down,mouse_over,mouse_out,
             hide: hide_all,
-            activate_and_highlight_selected
+            activate_and_highlight_selected,
         };
 
-        data.ports.borrow().init_frp(&network,port_frp);
+        data.ports.borrow().init_frp(&network_internal,port_frp);
+
+        *self.network_internal.borrow_mut() = network_internal;
 
         // // FIXME this is a hack to ensure the ports are invisible at startup.
         // // Right now we get some of FRP mouse events on startup that leave the
         // // ports visible by default.
         // // Once that is fixed, remove this line.
         delay_hide.from_now_to(TWEEN_END_VALUE);
 
-        self
+        self.set_port_colors_based_on_available_types()
     }
 
-    // TODO: Implement proper sorting and remove.
-    /// Hack function used to register the elements for the sorting purposes. To be removed.
-    pub(crate) fn order_hack(scene:&Scene) {
-        let logger = Logger::new("output shape order hack");
-        component::ShapeView::<multi_port_area::Shape>::new(&logger,scene);
-        component::ShapeView::<single_port_area::Shape>::new(&logger,scene);
+    fn set_port_colors_based_on_available_types(&self) {
+        self.id_map.borrow().iter().for_each(|(id, crumb)|{
+            let color = self.get_port_color(crumb).unwrap_or(DEFAULT_TYPE_COLOR);
+            self.data.ports.borrow().set_color(*id,color);
+        })
+    }
+
+    /// Return the color of the port indicated by the given `Crumb`.
+    pub fn get_port_color(&self, crumbs:&[span_tree::Crumb]) -> Option<color::Lcha> {
+        let ast_id = self.expression.borrow().get_id_for_crumbs(crumbs)?;
+        self.type_map.type_color(ast_id)
+    }
+
+    /// Set the type information for the given `ast::Id`.
+    pub fn set_expression_type(&self, id:ast::Id, maybe_type:Option<Type>) {
+        // TODO should we ignore this if it is not part f our expression?
+        match maybe_type {
+            Some(r#type) => self.type_map.insert(id,r#type),
```

Comment:
I don't think this is really the var name that we are looking. Even `tp` would be better than this strange monster ;)